internal/restic/snapshot.go

```diff
@@ -27,17 +27,45 @@ type Snapshot struct {
 	id *ID // plaintext ID, used during restore
 }
 
+func pathSplit(path string) (root string, lst []string) {
+	for root = filepath.Clean(path); ; {
+		var file string
+		root, file = filepath.Split(root)
+		if file == "" {
+			break
+		}
+		lst = append([]string{file}, lst...)
+	}
+	return
+}
+
+func pathMangle(path string, prefix string, strip int) string {
+	pathRoot, pathList := pathSplit(path)
+	if strip > 0 {
+		var stripRoot int
+		if pathRoot != "" {
+			stripRoot++
+		}
+		pathList = pathList[strip-stripRoot:]
+		pathRoot = ""
+	}
+	if prefix != "" {
+		pathRoot = prefix
+	} else {
+		absPath, err := filepath.Abs(pathRoot)
+		if err == nil {
+			pathRoot = absPath
+		}
+	}
+	return filepath.Join(append([]string{pathRoot}, pathList...)...)
+}
+
 // NewSnapshot returns an initialized snapshot struct for the current user and
 // time.
-func NewSnapshot(paths []string, tags []string, hostname string, time time.Time) (*Snapshot, error) {
+func NewSnapshot(paths []string, tags []string, hostname string, time time.Time, prefix string, strip int) (*Snapshot, error) {
 	absPaths := make([]string, 0, len(paths))
 	for _, path := range paths {
-		p, err := filepath.Abs(path)
```

Comment:
Hm, why did you remove the call to `filepath.Abs()`? That's probably wrong for non-absolute paths...

Or does `pathMangle` makes the paths absolute? That must be in the documentation for the function!

Reply:
Before my changes, if the Abs function returned an error, the function just appended the non-absolute path to the abs list:

```
                p, err := filepath.Abs(path)
                if err == nil {
                        absPaths = append(absPaths, p)
                } else {
                        absPaths = append(absPaths, path)
                }
```
The pathMangle is applied both to NewSnapshot and FindLatestSnapshot, since the behavior should be the same to determine actual target paths. For each path:
1) split each path into a root string (drive letter for Windows, slash for Unix, etc) and a list of path components with pathSplit();
2) if root string is empty, the path is relative, otherwise it's absolute;
3) if strip > 0 and the path is absolute, blank the root string and remove N-1 leading path elements; if strip > 0 and path is relative, remove N-1 leading path elements;
4) if prefix is not empty (either relative or absolute), it is applied in place of the root string;
5) if the prefix is empty, the root string is passed through filepath.Abs, so if a prefix is not applied, the output path is made absoluted, if possible. If filepath.Abs fails, the procedure continues with root string unchanged;
6) Finally the root string and the path elements are reassembled with filepath.Join.

So, to answer your question, pathMangle makes path absolute unless prefix is relative or unless filepath.Abs fails (as it was previously).

Reply:
Amending my previous statement, if strip > 0, path is not made absolute either.